File: #219_Contains_Duplicate_II/Solution.java
import java.util.*;

public class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        if (nums == null || nums.length < 2 || k == 0) return false;
        Map<Integer, List<Integer>> map = new HashMap();
        for (int i = 0; i < nums.length; i++) {
        	if (!map.containsKey(nums[i])) {
        		List<Integer> list = new ArrayList<Integer>();
        		list.add(i);
        		map.put(nums[i], list);
        	}
        	else {
        		map.get(nums[i]).add(i);
        	}
        }
        for (List<Integer> ls : map.values()) {
        	if (ls.size() < 2) continue;
        	for (int i = 1; i < ls.size(); i++) {
        		if (ls.get(i) - ls.get(i - 1) <= k) return true;
        	}
        }
        return false;
    }
}

File: #219_Contains_Duplicate_II/Solution.py
class Solution(object):
    def containsNearbyDuplicate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        if not nums:
        	return False
        if k <= 0:
        	return False
        window = set()
        i, j = 0, 0
        while j <= k and j < len(nums):
        	if nums[j] in window:
        		return True
        	window.add(nums[j])
        	j += 1
        while j < len(nums):
        	window.remove(nums[i])
        	i += 1
        	if nums[j] in window:
        		return True
        	window.add(nums[j])
        	j += 1
        return False

sol = Solution()
print sol.containsNearbyDuplicate([1, 2, 1, 4, 5], 2)